libbpf-cargo/src/btf/btf.rs

```diff
@@ -0,0 +1,739 @@
+use std::cmp::{max, min};
+use std::collections::BTreeSet;
+use std::convert::TryFrom;
+use std::ffi::{c_void, CStr, CString};
+use std::fmt::Write;
+use std::mem::size_of;
+use std::slice;
+
+use anyhow::{bail, ensure, Result};
+use scroll::Pread;
+
+use crate::btf::c_types::*;
+use crate::btf::*;
+
+pub struct Btf<'a> {
+    types: Vec<BtfType<'a>>,
+    ptr_size: u32,
+    string_table: &'a [u8],
+    bpf_obj: *mut libbpf_sys::bpf_object,
+}
+
+impl<'a> Btf<'a> {
+    pub fn new(name: &str, object_file: &[u8]) -> Result<Option<Self>> {
+        let cname = CString::new(name)?;
+        let mut obj_opts = libbpf_sys::bpf_object_open_opts::default();
+        obj_opts.sz = std::mem::size_of::<libbpf_sys::bpf_object_open_opts>() as libbpf_sys::size_t;
+        obj_opts.object_name = cname.as_ptr();
+        let bpf_obj = unsafe {
+            libbpf_sys::bpf_object__open_mem(
+                object_file.as_ptr() as *const c_void,
+                object_file.len() as u64,
+                &obj_opts,
+            )
+        };
+        if bpf_obj.is_null() {
```

Comment:
libbpf doesn't really return NULL on error, rather encoded error code as a pointer. You need to use libbpf_get_error() to check if it was an error or not.

Reply:
Good catch, thanks